src/DeviceInput/InputDevices/deviceSourceManager.ts

```diff
@@ -0,0 +1,308 @@
+        const deviceType = this._getDeviceTypeFromName(deviceName);
+
+        if (!this._devices[deviceType]) {
+            this._devices[deviceType] = new Array<DeviceSource<DeviceType>>();
+
+            if (deviceType == DeviceType.Touch) {
+                this._devices[deviceType][0] = new DeviceSource<DeviceType>(this._deviceInputSystem, POINTER_DEVICE, DeviceType.Touch, 0);
+            }
+        }
+
+        // If device is a touch device, update only touch points.  Otherwise, add new device.
+        if (deviceType == DeviceType.Touch) {
+            this._devices[deviceType][0].addTouchPoints(deviceName);
+        }
+        else {
+            this._devices[deviceType][deviceSlot] = new DeviceSource<DeviceType>(this._deviceInputSystem, deviceName, deviceType, deviceSlot);
+        }
+
+        this._updateFirstDevices(deviceType);
+    }
+
+    /**
+     * Function to remove device name to device list
+     * @param deviceName Name of Device
+     */
+    private _removeDevice(deviceName: string) {
+        const deviceSlot = this._getDeviceSlot(deviceName);
+        const deviceType = this._getDeviceTypeFromName(deviceName);
+
+        if (deviceType == DeviceType.Touch) {
+            this._devices[deviceType][0].removeTouchPoints(deviceName);
+        }
+        else {
+            delete this._devices[deviceType][deviceSlot];
+        }
+    }
+
+    /**
+     * Get slot for a given input
+     * @param deviceName Name of Device
+     * @returns Slot Number
+     */
+    private _getDeviceSlot(deviceName: string): number {
+        const splitName = deviceName.split("-");
+        const deviceSlot = parseInt(splitName[splitName.length - 1]);
+
+        if (deviceSlot) {
+            return deviceSlot;
+        }
+
+        return 0;
+    }
+
+    /**
+     * Updates array storing first connected device of each type
+     * @param type Type of Device
+     */
+    private _updateFirstDevices(type: DeviceType) {
+        switch (type) {
+            case DeviceType.Keyboard:
+            case DeviceType.Mouse:
+            case DeviceType.Touch:
+                this._firstDevice[type] = 0;
+                break;
+            case DeviceType.DualShock:
+            case DeviceType.Xbox:
+            case DeviceType.Switch:
+            case DeviceType.Generic:
+                let i = 0;
+                let first = -1;
+
+                while (first < 0 && i < this._devices[type].length) {
+                    if (this._devices[type][i]) {
+                        first = i;
+                    }
+                    i++;
+                }
+
+                this._firstDevice[type] = first;
+                break;
+        }
+    }
+
+    /**
+     * Gets DeviceType from the device name
+     * @param deviceName Name of Device from DeviceInputSystem
+     * @returns DeviceType enum value
+     */
+    private _getDeviceTypeFromName(deviceName: string): DeviceType {

```

Comment:
Moved to DeviceInputSystem and renamed to _getGamepadDeviceType.  This is because I removed the unnecessary if statements for Keyboard and Pointers.  The only thing that's left is gamepad stuff.  Additionally, the _getDeviceSlot function no longer exists.